### PR TITLE
[GC] Fix & optimize arrayEquals on AArch64 (#743)

### PR DESCRIPTION
**Pull request by weixlu**
[GC] Fix & optimize arrayEquals on AArch64 (#743)

Summary: fix tail comparison and simplify flow control for arrayEquals

Testing: jdk/javadoc/tool/parser/7091528/T7091528.java

Reviewers: mmyxym, linade

Issue: https://github.com/dragonwell-project/dragonwell11/issues/743

### CODE REVIEW

**Comment by weixlu** [comment]
About correctness 
This is to fix [https://github.com/dragonwell-project/dragonwell11/issues/743](url)
```
export test=test/langtools/jdk/javadoc/tool/parser/7091528/T7091528.java
function runJtreg() { jtreg -Xcomp -XX:+UseCompactObjectHeaders -ea -esa -timeoutFactor:4 -v:fail,error,time,nopass -nr -w $dir/index-$1 $test &> $dir/$1.log ; if [[ 0 -ne $? ]] ; then echo -n "$1 " ; else rm -rf $dir/index-$1 $dir/$1.log ; fi ; } ; export -f runJtreg ; export dir="tmp-jtreg-"`basename ${test##* } .java` ; rm -rf $dir ; mkdir -p $dir ; time seq 100 | xargs -i -n 1 -P `nproc` bash -c "runJtreg {}" ; echo total fail number: `ls $dir/*.log 2> /dev/null | wc | awk '{print $1}'`
```
with our fix, no failure is observed for five rounds of such test run.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 1]
```diff
@@ -5533,91 +5533,102 @@ address MacroAssembler::arrays_equals(Register a1, Register a2, Register tmp3,
       }
     }
   } else {
-    Label NEXT_DWORD, SHORT, TAIL, TAIL2, STUB,
-        CSET_EQ, LAST_CHECK;
+    Label NEXT_64BYTE, NEXT_16BYTE, SHORT, TAIL, TAIL03, TAIL01;
```
Comment (mmyxym) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:5536]: We'd better not change the default behavior. Make the new code under guard of UseCompactObjectHeaders.

Comment (weixlu) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:5536]: > We'd better not change the default behavior. Make the new code under guard of UseCompactObjectHeaders.

done. and also add simd check.

**Comment by weixlu** [comment]
@mmyxym Reviewer proposed to replace the tail check to the end, and we have tested its performance.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 2]
```diff
@@ -5532,6 +5532,107 @@ address MacroAssembler::arrays_equals(Register a1, Register a2, Register tmp3,
         cbnzw(tmp5, DONE);
       }
     }
+  } else if (UseCompactObjectHeaders) {
+    Label NEXT_64BYTE, NEXT_16BYTE, TAIL, TAIL03, TAIL01;
+    mov(result, false);
+    // if (a1 == null || a2 == null)
+    //     return false;
```
Comment (linade) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:5539]: Is there a reference implementation for this null check?

One of the existing implementation uses:
    tst(a1, a2);
which I think is quite clever; the other one uses:
    cbz(a1, DONE);
    ldrw(cnt1, Address(a1, length_offset));
    cbz(a2, DONE);
    ldrw(cnt2, Address(a2, length_offset));
    cmpw(cnt2, cnt1);
which moves ldrw(cnt1) ahead. My guess is that cmpw depends on the availability of cnt1 so it wants to load cnt1 sooner. Is there any evaluation that suggests the proposed implementation is superior?
    cbz(a1, DONE);
    cbz(a2, DONE);
    ldrw(cnt1, Address(a1, length_offset));
    ldrw(cnt2, Address(a2, length_offset));
    cmpw(cnt2, cnt1);

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 3]
```diff
@@ -5532,6 +5532,107 @@ address MacroAssembler::arrays_equals(Register a1, Register a2, Register tmp3,
...
+    ld1(v0, v1, v2, v3, T16B, Address(post(a1, wordSize * 8)));
+    ld1(v4, v5, v6, v7, T16B, Address(post(a2, wordSize * 8)));
+    sub(cnt1, cnt1, 8 * elem_per_word);
+    eor(v0, T16B, v0, v4);
+    eor(v1, T16B, v1, v5);
+    eor(v2, T16B, v2, v6);
+    eor(v3, T16B, v3, v7);
+    orr(v0, T16B, v0, v1);
+    orr(v0, T16B, v0, v2);
+    orr(v0, T16B, v0, v3);
```
Comment (linade) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:5601]:
```
    orr(v0, T16B, v0, v1);
    orr(v0, T16B, v0, v2);
    orr(v0, T16B, v0, v3);
```
changed to
```
    orr(v0, T16B, v0, v1);
    orr(v2, T16B, v2, v3);
    orr(v0, T16B, v0, v2);
```
might be better?

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 4]
```diff
@@ -5532,6 +5532,107 @@ address MacroAssembler::arrays_equals(Register a1, Register a2, Register tmp3,
...
+    cmpw(cnt2, cnt1);
+    br(NE, DONE);
+    // if (a1.length == 0)
+    //      return true;
+    cbz(cnt1, SAME);
+    lea(a1, Address(a1, base_offset));
+    lea(a2, Address(a2, base_offset));
+
+    cmpw(cnt1, elem_per_word);
+    if (UseSIMDForArrayEquals) {
```
Comment (linade) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:5555]: highly-probable case ([8, 64) bytes) needs two jumps, need reordering

the default implementation only requires two fall-throughs

**Review by mmyxym — APPROVED** [review]

**Review by linade — APPROVED** [review]
Looks good as a fix for the compressed header issue. It might need more comprehensive benchmarking before considering enabling by default.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp (hunk 4 above)
Comment (weixlu) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:5556]: done

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp (hunk 2 above)
Comment (weixlu) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:5539]: done

**Pull request merged**